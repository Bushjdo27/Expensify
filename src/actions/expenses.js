import {
    ADD_EXPENSE,
    ALL_EXPENSE,
    UPDATE_EXPENSE,
    DELETE_EXPENSE
} from './CONSTANT';


export const addExpense = ({id,description, note, amount, createdAt }) => {
    return {
        type: ADD_EXPENSE,
        expense: {
            id,
            description,
            note,
            amount,
            createdAt
        }
        

    }
}

export const updateExpense = (id, update) => {
    return {
        type: UPDATE_EXPENSE,
        id,
        expense: update

    }
}

export const deleteExpense = (id) => {
    return {
        type: DELETE_EXPENSE,
        id

    }
}