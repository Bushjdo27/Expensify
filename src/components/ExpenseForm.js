import React, {Component} from 'react';
//import { connect } from 'react-redux';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates'

class ExpenseForm extends Component{

    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            err: false
        }
    }

    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=>{
            return {
                description
            }
        })
    }

    onAmountChange = (e)=>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(.\d{0,2})?$/)){
            this.setState(() => ({amount}))
        }
    }

    onDateChange = (createdAt) => {
        this.setState(()=>({createdAt}))
    }
    onFocusChange = ({focused}) =>{
        this.setState(()=>({calendarFocused: focused}))
    }

    onNoteChange = (e)=>{
        const note = e.target.value;
        this.setState(() => ({note}))
    }

    onSubmit = (e)=>{
        e.preventDefault();
        const {description, amount, note,createdAt} = this.state;
        if(description && amount){
            this.setState(()=>({err: false}))
            if(this.props.edit){
                this.props.onSubmit({
                    description,
                    amount: parseFloat(amount),
                    note,
                    createdAt: createdAt.valueOf()
                })
            }else{
                this.props.onSubmit({
                    id: "3",
                    description,
                    amount: parseFloat(amount),
                    note,
                    createdAt: createdAt.valueOf()
                })
            }
            
            
        }else {
            this.setState(()=>({err: true}))
        }
        
    }
    render(){
        const {description , amount , note, createdAt,calendarFocused,err} = this.state;
        return (
            <div>
                <form className="form" onSubmit={this.onSubmit}>
                    <div className="input__box">
                        <input 
                            onChange={this.onDescriptionChange} 
                            value={description} 
                            type="text" 
                            id="description" 
                            name="description" 
                            required />
                        <label htmlFor="description"> Description</label>
                    </div>

                    <div className="input__box">
                        <input 
                            onChange={this.onAmountChange} 
                            value={amount} 
                            type="text" 
                            id="amount" 
                            name="amount" 
                            required />
                        <label htmlFor="amount"> Amount</label>
                    </div>
                    <div className="input__box">
                        <SingleDatePicker 
                            date={createdAt}
                            onDateChange={this.onDateChange}
                            focused={calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={()=>false}
                        />
                    </div>

                    

                    <div className="input__box">
                        <textarea
                            onChange={this.onNoteChange}
                            value={note}
                            name="note"
                            rows="6" 
                            placeholder="Add a note for your expense (optional)"
                        />
                    </div>

                    <div className="u--center bg-white u--medium">
                        {err && <p>Please enter description, amount for your expense</p>}
                        <button className="btn btn--submit">{this.props.edit ? 'Edit' : 'Add'}</button>
                    </div>
                    
                </form>
            </div>
        )
    }
}


//export default connect()(ExpenseForm);
export default ExpenseForm;