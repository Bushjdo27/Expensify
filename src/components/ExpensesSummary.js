import React from 'react';
import numeral from 'numeral'

export const ExpensesSummary = ({expenses = []}) => {
    const totalling = expenses.reduce((accumulateValue, currentValue) => {
                            return accumulateValue += currentValue.amount
                        },0);
    const viewing = expenses.length;
    return (
        <div>
            <p>Viewing : {viewing}</p>
            <p>
                Totalling : {
                    numeral(totalling).format('$0,0.00')
                }
            </p>
        </div>
    )
}
