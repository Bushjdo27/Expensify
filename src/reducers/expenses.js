import {
    ADD_EXPENSE,
    ALL_EXPENSE,
    UPDATE_EXPENSE,
    DELETE_EXPENSE
} from '../actions/CONSTANT';


export default function(state = [], action){

    switch(action.type) {
        case ALL_EXPENSE: 
            return state;
        case ADD_EXPENSE: 
            return [
                ...state,
                action.expense
            ];
        case UPDATE_EXPENSE: 
            return state.map(expense => {
                if(expense.id === action.expense.id){
                    return action.expense
                }
                return expense
            });
        case DELETE_EXPENSE: 
            return state.filter(expense => expense.id !== action.id);
        
        default: 
            return state;
    }
}