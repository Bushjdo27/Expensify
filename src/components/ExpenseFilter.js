import React from 'react';
import { connect } from 'react-redux';
import {setTextFilter, sortByAmount , sortByDate, setStartDate, setEndDate} from '../actions/filters';
import { DateRangePicker} from 'react-dates'

export class ExpenseFilter extends React.Component {

    state = {
        calendarFocus: null,
    }

    onDateChange = ({startDate, endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calendarFocus)=>{
        this.setState(()=>({calendarFocus}))
    }
    onSelectChange = (e) => {
        if(e.target.value === 'date'){
            this.props.sortByDate()
        }else if(e.target.value === 'amount'){
            this.props.sortByAmount()
        }
    }
    onTextChange = (e)=>{
        const text = e.target.value;
        this.props.setTextFilter(text)
    }
    render(){
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                    />
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
    
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDateChange}
                    focusedInput={this.state.calendarFocus}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                />
            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (text)=> dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseFilter);