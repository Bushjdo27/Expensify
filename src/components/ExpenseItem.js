import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import {connect} from 'react-redux';
import {deleteExpense} from '../actions/expenses';
import { Link } from 'react-router-dom';


export const ExpenseItem = ({id, dispatch, description, amount, createdAt}) => {
    return (
        <div>
            <Link className="link-text" to={`/edit/${id}`}>
                <h3>{description}</h3>
            </Link>
            <p>
                <strong>{numeral(amount).format('$0,0.00')} </strong>  
                    - 
                {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
            <button 
                onClick={()=> {
                    dispatch(deleteExpense(id))
                }}>
                Remove
            </button>
        </div>
    )
}

export default connect()(ExpenseItem);