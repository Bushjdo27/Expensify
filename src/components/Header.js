import React , {Component} from 'react';
import { NavLink, withRouter } from 'react-router-dom'
import { connect } from 'react-redux';
import { startLogOut } from '../actions/auth';

export class Header extends Component{

    handleSignOut = ()=>{
        this.props.startLogOut()
    }
    render() {
        return (
            <header className="header">
                <div className="header__logo">
                    <h1>Expensify</h1>
                </div>
                <nav>
                    <ul className="header__navigation">
                        <li className="header__navigation__item">
                            <NavLink className="header__navigation__item--link" to="/dashboard" activeClassName="active-nav">
                                Dashboard
                            </NavLink>
                        </li>
                        <li className="header__navigation__item">
                            <NavLink className="header__navigation__item--link" to="/create" activeClassName="active-nav">
                                Create
                            </NavLink>
                        </li>
                        <li className="header__navigation__item">
                            <button onClick={this.handleSignOut} className="header__navigation__item--link">
                                Log Out
                            </button>
                        </li>
                        
                    </ul>
                </nav>
            </header>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startLogOut: () => dispatch(startLogOut())
    }
}

export default withRouter(connect(undefined, mapDispatchToProps)(Header));