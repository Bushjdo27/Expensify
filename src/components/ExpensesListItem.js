import React from 'react';
import ExpenseItem from './ExpenseItem'
import selectExpenses from '../selector/expenses';
import {ExpensesSummary} from './ExpensesSummary';
import {connect} from 'react-redux';

export const ExpensesListItem = (props) => {
     return (
         <div>
            {props.expenses.length === 0 ?
                <p>No Expense</p> : 
                props.expenses.map(expense => {
                    return <ExpenseItem key={expense.id} {...expense}/>
                })
            }
            <ExpensesSummary expenses={props.expenses}/>
         </div>
     )
}



const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses,state.filters)
    }
}


export default connect(mapStateToProps)(ExpensesListItem);