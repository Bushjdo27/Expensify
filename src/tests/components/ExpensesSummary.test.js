import React from 'react';
import {shallow} from 'enzyme';
import numeral from 'numeral';
import expenses from '../fixtures/expenses'
import {ExpensesSummary} from '../../components/ExpensesSummary';

test('should render ExpensesSummary correctly',()=>{
    const wrapper = shallow(<ExpensesSummary expenses={[]}/>);

    expect(wrapper).toMatchSnapshot();
})

test('should render with no props expenses', ()=>{
    const wrapper = shallow(<ExpensesSummary />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').at(0).text()).toBe("Viewing : 0");
    expect(wrapper.find('p').at(1).text()).toBe(`Totalling : ${numeral(0).format('$0,0.00')}`);
});

test('should render with expenses list', ()=>{
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
    const total = expenses.reduce((accum, current)=>{
        return accum += current.amount
    },0)
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').at(0).text()).toBe(`Viewing : ${expenses.length}`);
    expect(wrapper.find('p').at(1).text()).toBe(`Totalling : ${numeral(total).format('$0,0.00')}`);

})