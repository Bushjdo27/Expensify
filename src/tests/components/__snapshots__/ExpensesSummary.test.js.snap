// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  <p>
    Viewing : 
    0
  </p>
  <p>
    Totalling : 
    $0.00
  </p>
</div>
`;

exports[`should render with expenses list 1`] = `
<div>
  <p>
    Viewing : 
    3
  </p>
  <p>
    Totalling : 
    $114,195.00
  </p>
</div>
`;

exports[`should render with no props expenses 1`] = `
<div>
  <p>
    Viewing : 
    0
  </p>
  <p>
    Totalling : 
    $0.00
  </p>
</div>
`;
